### DISCUSSION TITLE
Timestamps while working with code

### DISCUSSION THREAD

**Discussion by hussamsayeed** [openai/whisper#453] [General]
How do I get the timestamps while working with the code? I see timestamps are visible while working with command.
![Screenshot (798)](https://user-images.githubusercontent.com/46478199/199445326-b5ce2858-93d6-42da-9381-d688e829eda3.png)

**Comment by levaweb** [comment]
I've seen in the  examples the keyword: without_timestamps=True/False
Can we use it to get the timestamp printed out when executing the code in Python?
Or, how do we get the timestamp in a separate file?
It looks weird that timestamps are only showing up in the CMD prompt mode.
Thanks

**Comment by jongwook** [comment]
They are returned in the field `result["segments"]`.

**Comment by hussamsayeed** [comment]
since the output is in dictionary format, we have to open it as iterator, you can add this code to get the time stamps.

```
import whisper
import zlib
from typing import Iterator, TextIO
```

```
def format_timestamp(seconds: float, always_include_hours: bool = False, decimal_marker: str = '.'):
    assert seconds >= 0, "non-negative timestamp expected"
    milliseconds = round(seconds * 1000.0)

    hours = milliseconds // 3_600_000
    milliseconds -= hours * 3_600_000

    minutes = milliseconds // 60_000
    milliseconds -= minutes * 60_000

    seconds = milliseconds // 1_000
    milliseconds -= seconds * 1_000

    hours_marker = f"{hours:02d}:" if always_include_hours or hours > 0 else ""
    return f"{hours_marker}{minutes:02d}:{seconds:02d}{decimal_marker}{milliseconds:03d}"
```
```
def transcript_timestamp(transcript: Iterator[dict]):
    for segment in transcript:
        print(
            f"{format_timestamp(segment['start'])} --> {format_timestamp(segment['end'])} "
            f"Text: {segment['text'].strip().replace('-->', '->')}\n",
            flush=True
        )
```

Call the function by feeding the output

`trans=transcript_timestamp(result["segments"])`

**Reply by levaweb** [reply]
Thanks Hussam! Really appreciated

Is there a way to improve the _accuracy_ of the timestamps?
On average they looks +/-250 ms inaccurate.
Is this perhaps something due to rounding floats into integers?

Thanks

**Reply by hussamsayeed** [reply]
Check with this it may works #435

**Comment by levaweb** [comment]
Awesome! Thanks for sharing

**Comment by tflax32** [comment]
Here is a way to get the srt file from your whisper transcription using the pysrt library:
```
import whisper
import pysrt

model = whisper.load_model("base")
audio_file = ("YOUR AUDIO FILE PATH")
result = model.transcribe(audio_file, language='en')

subs = pysrt.SubRipFile()
sub_idx = 1

for i in range(len(result["segments"])):
    start_time = result["segments"][i]["start"]
    end_time = result["segments"][i]["end"]
    duration = end_time - start_time
    timestamp = f"{start_time:.3f} - {end_time:.3f}"
    text = result["segments"][i]["text"]
    
    sub = pysrt.SubRipItem(index=sub_idx, start=pysrt.SubRipTime(seconds=start_time), 
                           end=pysrt.SubRipTime(seconds=end_time), text=text)
    subs.append(sub)
    sub_idx += 1
    
subs.save("output.srt")
```
